thunderfish/psdtype.py

```diff
@@ -0,0 +1,173 @@
+import numpy as np
+import sys
+
+def bin_it(power, freq_bins, max_freq, res):
+    """
+    Take a 1-D array of powers (from powerspectrums), transforms it into dB and devides it into several bins.
+
+    :param power:           (1-D array) power array of a psd.
+    :param freq_bins:       (float) width of frequency bins in which the psd shall be divided (Hz).
+    :param max_freq:        (float) maximum frequency that shall be provided in the separated power array.
+    :param res:             (float) the frequency resolution of the power array comming from a psd.
+                            res = freqs[-1]/len(freqs)
+    :return power_db:       (2-D array) list of lists with the in bins divided powers of the powerspectrum transforemed
+                            into dB.
+    """
+    power_db = []
+    for trial in np.arange(max_freq / freq_bins):
+        tmp_power_db = 10.0 * np.log10(power[trial * int( freq_bins / (res) ) : (trial +1) * int( freq_bins / (res) ) - 1])
+        power_db.append(tmp_power_db)
+    return power_db
+
+def get_bin_percentiles(power_db):
+    """
+    Takes a 2-D array if lists. For every list the function calculates several percentails that are later returned.
+
+    :param power_db:        (2-D array) list of lists with the in bins divided powers of the powerspectrum transforemed
+                            into dB.
+    :return:                (2-D array) for every bin four values are calulated and stored in separate lists. These four
+                            values are percentiles of the respective bins.
+    """
+    power_db_top = np.ones(len(power_db))
+    power_db_upper_middle = np.ones(len(power_db))
+    power_db_lower_middle = np.ones(len(power_db))
+    power_db_bottom = np.ones(len(power_db))
+
+    for fbin in np.arange(len(power_db)):
+        power_db_top[fbin] = np.percentile(power_db[fbin], 99)
```

Comment:
the actual percentiles MUST be arguments of the function!

You can use the percentile function with a  list of percentiles. percentile sorts the data and then returns the percentile. Calling percentile once with a list of percentiles sorts the data only once. Sorting is computationally expensive.

a = 1
b= 25
np.percentile(power_db[fbin], [a, b, 100-b, 100-a])